File: src/core/modules/Math/Math.cpp
#include "Math.h"
#include <cmath>
#include <numbers>

namespace core::modules::Math {
	namespace {
		bool Initialized = false;
        const double DEG2RAD_FACTOR = std::numbers::pi / 180.0;
        const double RAD2DEG_FACTOR = 180.0 / std::numbers::pi;
	}
    bool Init() {
        if (Initialized) {
            return true;
        }
        Initialized = true;
        return true;
    }
    double sin(double deg) {
        return std::sin(deg * DEG2RAD_FACTOR);
    }
    double cos(double deg) {
        return std::cos(deg * DEG2RAD_FACTOR);
    }
    double tan(double deg) {
        return std::tan(deg * DEG2RAD_FACTOR);
    }
    double atan2(double dy, double dx) {
        return std::atan2(dy, dx) * RAD2DEG_FACTOR;
    }
    double deg2rad(double deg) {
        return deg * DEG2RAD_FACTOR;
    }
    double rad2deg(double rad) {
        return rad * RAD2DEG_FACTOR;
    }
    void Shutdown() {
        if (!Initialized) {
            return;
        }
        Initialized = false;
    }
}

File: doc/modules/Math.lua
---@meta
modules = modules or {}

---@class modules.Math
modules.Math = {}

---计算角度的正弦值。
---参数使用角度制，而不是弧度制。
---@param deg number 角度
---@return number value 正弦值
function modules.Math.sin(deg) end

---计算角度的余弦值。
---参数使用角度制，而不是弧度制。
---@param deg number 角度
---@return number value 余弦值
function modules.Math.cos(deg) end

---计算角度的正切值。
---参数使用角度制，而不是弧度制。
---@param deg number 角度
---@return number value 正切值
function modules.Math.tan(deg) end

---根据二维向量 (dx, dy) 计算方向角。
---返回值使用角度制，通常位于 [-180, 180] 范围内。
---@param dy number Y 方向差值
---@param dx number X 方向差值
---@return number deg 方向角（角度）
function modules.Math.atan2(dy, dx) end

---将角度转换为弧度。
---@param deg number 角度
---@return number rad 弧度
function modules.Math.deg2rad(deg) end

---将弧度转换为角度。
---@param rad number 弧度
---@return number deg 角度
function modules.Math.rad2deg(rad) end
